### DISCUSSION TITLE
signed load/store/extract vs explicit conversion instructions

### DISCUSSION THREAD

**Discussion by magenbluten** [mrisc32/mrisc32#133] [General]
the instruction set has signed and unsigned load store instructions and also (bit) extraction instructions. i was wondering why having those and not just use explicit conversion instructions like sign-extend or zero-extend (could use the bit-offset to do the conversion instead of byte offsets). is there a reason to favor one or another? (a signed load would then be 2 instructions of course)

**Comment by mbitsnbites** [comment]
The main reason is that, as you say, signed load would be two instructions (or unsigned load, depending on what you select as your default).

Loading data from memory is a very common operation, and requiring two instructions for that would hurt both code density and performance. Besides, dynamically selecting sign-extension or zero-extension in the memory load unit is trivial, and only two additional instructions are required (`LDUH`, `LDUB`).

**Edit:** There are actually two ways of doing sign-extension of 8-bit and 16-bit values: Either with the `EBF` instruction or the `SHUF` instruction. Same thing with zero-extension (`EBFU` or `SHUF`).
